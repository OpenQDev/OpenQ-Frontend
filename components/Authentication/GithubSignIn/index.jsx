// Third party
import React from 'react';
import { useRouter } from 'next/router';
// Custom
import Image from 'next/image';

const GithubSignIn = ({ redirectUrl, signInStyle }) => {
  const router = useRouter();

  const signIn = () => {
    const clientId = `client_id=${process.env.NEXT_PUBLIC_OPENQ_ID}`;
    const nonce = randomString(10);
    window.localStorage.setItem('csrf_nonce', nonce);
    const state = {
      [nonce]: {
        redirectUrl,
      },
    };
    const stateParams = `state=${JSON.stringify(state)}`;
    router.push(
      `https://github.com/login/oauth/authorize?${clientId}&${stateParams}&scope=read:user%20read:org%20public_repo`
    );
  };

  function randomString(length) {
    return Array(length + 1)
      .join((Math.random().toString(36) + '00000000000000000').slice(2, 18))
      .slice(0, length);
  }

  return (
    <button
      onClick={() => signIn()}
      className={`flex justify-center ${signInStyle ? signInStyle : 'btn-default'} whitespace-nowrap w-full`}
    >
      <div className='flex flex-row items-center justify-center space-x-3'>
        <div className='h-4 w-4 md:h-6 md:w-6 relative'>
          <Image src='/social-icons/github-logo-white.svg' alt='Picture of the author' width={24} height={24} />
        </div>
        <div>Sign In</div>
      </div>
    </button>
  );
};

export default GithubSignIn;
