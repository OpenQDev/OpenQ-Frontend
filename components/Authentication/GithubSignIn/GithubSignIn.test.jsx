import React from 'react';
import GithubSignIn from '.';
import renderer from 'react-test-renderer';
import nextRouter from 'next/router';

nextRouter.useRouter = vi.fn();
nextRouter.useRouter.mockImplementation(() => ({
  query: { type: null },

  prefetch: vi.fn(() => {
    return { catch: vi.fn };
  }),
}));

describe('GithubSignIn Snapshot', () => {
  it('should match DOM Snapshot', () => {
    const tree = renderer.create(<GithubSignIn />);
    expect(tree.toJSON()).toMatchSnapshot();
  });
});
