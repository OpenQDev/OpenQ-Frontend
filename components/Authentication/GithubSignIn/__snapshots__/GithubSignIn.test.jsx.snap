// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`GithubSignIn Snapshot > should match DOM Snapshot 1`] = `
<button
  className="flex justify-center btn-default whitespace-nowrap w-full"
  onClick={[Function]}
>
  <div
    className="flex flex-row items-center justify-center space-x-3"
  >
    <div
      className="h-4 w-4 md:h-6 md:w-6 relative"
    >
      <img
        alt="Picture of the author"
        data-nimg="1"
        decoding="async"
        height={24}
        loading="lazy"
        onError={[Function]}
        onLoad={[Function]}
        src="/social-icons/github-logo-white.svg"
        style={
          {
            "color": "transparent",
          }
        }
        width={24}
      />
    </div>
    <div>
      Sign In
    </div>
  </div>
</button>
`;
